Fix torch.compile with FakeTensor that has SymInt sizes

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107662

**Motivation:**
When input FakeTensor to torch.compile has SymInt sizes (e.g. make_fx(opt_f, tracing_mode="symbolic"):
1. We cannot create a FakeTensor from that input in dynamo due to the SymInts.
2. We cannot check input tensors in guard check function and will abort due to tensor check calls sizes/strides.

For 1, we specialize the FakeTensor's SymInts using their hints. This is mostly safe since inputs mostly have concrete shapes and not computed from some DynamicOutputShape ops. We'll throw a data dependent error if the symint is unbacked. 

For 2, we replace size/stride calls with the sym_* variants in TENSOR_CHECK guards' check function.

**Test Plan:**
See added tests.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

ydwu4 committed:
Fix torch.compile with FakeTensor that has SymInt sizes

[ghstack-poisoned]

ydwu4 committed:
Update on "Fix torch.compile with FakeTensor that has SymInt sizes"


**Motivation:**
When input FakeTensor to torch.compile has SymInt sizes:
1. We cannot create a FakeTensor from that input in dynamo due to the SymInts.
2. We cannot check input tensors in guard check function and will abort due to tensor check calls sizes/strides.

For 1, we specialize the FakeTensor's SymInts using their hints. This is mostly safe since inputs mostly have concrete shapes and not computed from some DynamicOutputShape ops.

For 2, we replace size/stride calls to the sym_* variants.

**Test Plan:**
See added tests.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 committed:
Update on "Fix torch.compile with FakeTensor that has SymInt sizes"


**Motivation:**
When input FakeTensor to torch.compile has SymInt sizes:
1. We cannot create a FakeTensor from that input in dynamo due to the SymInts.
2. We cannot check input tensors in guard check function and will abort due to tensor check calls sizes/strides.

For 1, we specialize the FakeTensor's SymInts using their hints. This is mostly safe since inputs mostly have concrete shapes and not computed from some DynamicOutputShape ops.

For 2, we replace size/stride calls to the sym_* variants.

**Test Plan:**
See added tests.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 committed:
Update on "Fix torch.compile with FakeTensor that has SymInt sizes"


**Motivation:**
When input FakeTensor to torch.compile has SymInt sizes:
1. We cannot create a FakeTensor from that input in dynamo due to the SymInts.
2. We cannot check input tensors in guard check function and will abort due to tensor check calls sizes/strides.

For 1, we specialize the FakeTensor's SymInts using their hints. This is mostly safe since inputs mostly have concrete shapes and not computed from some DynamicOutputShape ops.

For 2, we replace size/stride calls to the sym_* variants.

**Test Plan:**
See added tests.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2235,6 +2244,20 @@ def create_symbolic_sizes_strides_storage_offset(
+            assert isinstance(maybe_sym, (int, torch.SymInt))
+            return maybe_sym if isinstance(maybe_sym, int) else maybe_sym.node.require_hint()
+
+        ex_size = pytree.tree_map(maybe_specialize_sym_int_with_hint, tuple(ex.size()))
```
There isn't any reason to do a tree_map here, just do a list comprehension for speed.

ezyang commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -54,8 +54,8 @@ class TensorCheck {
-    const auto& sizes = v.sizes();
-    const auto& strides = v.strides();
+    const auto& sizes = v.sym_sizes();
+    const auto& strides = v.sym_strides();
```
🤔  This is the right thing to do but it may make guard perf worse. Hopefully not too much worse.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2212,11 +2213,19 @@ def _produce_dyn_sizes(self,
+                                          dynamic_dims: DimList[DimDynamic],
+                                          constraint_dims: List[DimConstraint]
+                                          ) -> List[sympy.Expr]:
+        assert all(isinstance(val, int) for val in tensor_size), f"Expect size to be a plain tuple of ints but got {tensor_size}"
```
???

ydwu4 replied:
Is it because the added _produce_dyn_sizes_from_int_tuple and the redirection of _produce_dyn_sizes to it, or the assertions?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2235,6 +2244,20 @@ def create_symbolic_sizes_strides_storage_offset(
+        # Dynamo may want to wrap FakeTensors with SymInt sizes up e.g. make_fx(opt_f(), tracing_mode="symbolic").
+        # We create symbols in shape_env using the backed hints behind SymInt.
+        # Case 1: when SymInt is backed, dynamo can proceed with FakeTensors that have concrete shape.
+        # produce_guards will trigger specializations on the outer stuff
```
You probably want to talk us through the guard ordering problem here too

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2235,6 +2244,20 @@ def create_symbolic_sizes_strides_storage_offset(
+        # Case 2: when the SymInt is unbacked, we will throw an data dependent error in require_hint().
+        def maybe_specialize_sym_int_with_hint(maybe_sym) -> int:
+            assert isinstance(maybe_sym, (int, torch.SymInt))
+            return maybe_sym if isinstance(maybe_sym, int) else maybe_sym.node.require_hint()
```
Can you add an assert that maybe_sym.node.shape_env is not self?

ezyang reviewed (approved):
cool it actually worked lol

ydwu4 committed:
Update on "Fix torch.compile with FakeTensor that has SymInt sizes"


**Motivation:**
When input FakeTensor to torch.compile has SymInt sizes (e.g. make_fx(opt_f, tracing_mode="symbolic"):
1. We cannot create a FakeTensor from that input in dynamo due to the SymInts.
2. We cannot check input tensors in guard check function and will abort due to tensor check calls sizes/strides.

For 1, we specialize the FakeTensor's SymInts using their hints. This is mostly safe since inputs mostly have concrete shapes and not computed from some DynamicOutputShape ops. We'll throw a data dependent error if the symint is unbacked. 

For 2, we replace size/stride calls with the sym_* variants in TENSOR_CHECK guards' check function.

**Test Plan:**
See added tests.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov 

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
